Studies/source/GenStudy.cxx

```diff
@@ -0,0 +1,173 @@
+            const GenParticle* h_bb = nullptr;
+
+            for(const auto& higgs : higgsPair){
+                if(higgs->daughters.size() != 2)
+                throw analysis::exception("Higgs decay products must be 2, insteed has a size of '%1%'") %higgs->daughters.size();
+
+                if(std::abs(higgs->daughters.at(0)->pdg) == particles::ParticleCode::tau &&
+                    std::abs(higgs->daughters.at(1)->pdg) == particles::ParticleCode::tau)
+                    h_tautau = higgs;
+                else if(std::abs(higgs->daughters.at(0)->pdg) == particles::ParticleCode::b &&
+                    std::abs(higgs->daughters.at(1)->pdg) == particles::ParticleCode::b)
+                     h_bb = higgs;
+                 else
+                     throw analysis::exception("Higgs pair doesn't decay in a pair of taus or b jets h1 -> '%1%', h2 -> '%2%'.") %higgs->daughters.at(0)->pdg %higgs->daughters.at(1)->pdg;
+            }
+            if(h_tautau == nullptr || h_bb == nullptr)
+                throw analysis::exception("The Higgs pairs are not filled");
+            if(h_tautau->daughters.size() != 2 || h_bb->daughters.size() != 2)
+                throw analysis::exception("Each of the Higgs pairs must decay in two particles.");
+
+            int n_tau_matches = 0;
+            for (size_t tau_index = 0; tau_index < h_tautau->daughters.size(); tau_index++) {
+                std::vector<const GenParticle*> visible_daughters;
+                if(std::abs(h_tautau->daughters.at(tau_index)->momentum.eta()) > 2.3 ) continue;
+                auto visibleMomentum = genEvent.GetFinalStateMomentum(*h_tautau->daughters.at(tau_index), visible_daughters, true, false);
+                for (size_t reco_tau_index = 0; reco_tau_index < 2; reco_tau_index++) { 
+                    if(ROOT::Math::VectorUtil::DeltaR(visibleMomentum, event.lep_p4.at(reco_tau_index)) < deltaR_value)
+                        ++n_tau_matches;
+                }
+            }
+
+            const auto baryons_plus_mesons = genEvent.GetTypesParticles({"baryon", "meson"}, h_bb);
+            if(baryons_plus_mesons.size() < 2)
+                throw analysis::exception("There're less than two barions or mesons");
+
+            size_t max_i = 0;
+            std::vector<const GenParticle*> most_energetic_pair;
+            for (size_t bm_index = 0; bm_index < baryons_plus_mesons.size(); bm_index++) {
+                if(baryons_plus_mesons.at(bm_index)->momentum.E() > baryons_plus_mesons.at(max_i)->momentum.E())
+                    max_i = bm_index;
+            }
+
+            size_t second_max_i = max_i == 0 ? 1 : 0;
+            for (size_t bm_index = 0; bm_index < baryons_plus_mesons.size(); bm_index++) {
+                if(bm_index == max_i) continue;
+                if(baryons_plus_mesons.at(bm_index)->momentum.E() > baryons_plus_mesons.at(second_max_i)->momentum.E())
+                    second_max_i = bm_index;
+            }
```

Comment:
this algorithm is wrong. to have a more error-prone code, it is better to sort the indices.